compiler/shared_ast/expr.ml

```diff
@@ -756,3 +718,99 @@ let rec size : type a. (a, 't) gexpr -> int =
   | EEnumInj (e1, _, _) -> 1 + size e1
   | EMatchS (e1, _, cases) ->
     EnumConstructorMap.fold (fun _ e acc -> acc + 1 + size e) cases (size e1)
+
+(* - Expression building helpers - *)
+
+let make_var (x, mark) =
+  Bindlib.box_apply (fun x -> x, mark) (Bindlib.box_var x)
+
+let make_abs xs e taus pos =
+  let mark =
+    map_mark
+      (fun _ -> pos)
+      (fun ety ->
+        List.fold_right
+          (fun tx acc -> Marked.mark pos (TArrow (tx, acc)))
+          taus ety)
+      (Marked.get_mark (Bindlib.unbox e))
+  in
+  Bindlib.box_apply (fun b -> EAbs (b, taus), mark) (Bindlib.bind_mvar xs e)
+
+let make_app e u pos =
+  Bindlib.box_apply2
+    (fun e u ->
+      let mark =
+        fold_marks
+          (fun _ -> pos)
+          (function
+            | [] -> assert false
+            | fty :: argtys ->
+              List.fold_left
+                (fun tf tx ->
+                  match Marked.unmark tf with
+                  | TArrow (tx', tr) ->
+                    assert (unifiable tx.ty tx');
+                    (* wrong arg type *)
+                    tr
+                  | TAny -> tf
+                  | _ ->
+                    Format.eprintf
+                      "Attempt to construct application of non-arrow type %a:@\n\
+                       %a"
+                      Print.typ_debug tf
+                      (Print.expr_debug ~debug:false)
+                      e;
+                    assert false)
+                fty.ty argtys)
+          (List.map Marked.get_mark (e :: u))
+      in
+      EApp (e, u), mark)
+    e (Bindlib.box_list u)
+
+let empty_thunked_term mark =
+  let silent = Var.make "_" in
+  let pos = mark_pos mark in
+  make_abs [| silent |]
+    (Bindlib.box (ELit LEmptyError, mark))
+    [TLit TUnit, pos]
+    pos
+
+let make_let_in x tau e1 e2 mpos =
+  make_app (make_abs [| x |] e2 [tau] mpos) [e1] (pos (Bindlib.unbox e2))
+
+let make_multiple_let_in xs taus e1s e2 mpos =
+  make_app (make_abs xs e2 taus mpos) e1s (pos (Bindlib.unbox e2))
+
+let make_default exceptions just cons mark =
+  let rec bool_value = function
+    | ELit (LBool b), _ -> Some b
+    | EApp ((EOp (Unop (Log (l, _))), _), [e]), _
+      when l <> PosRecordIfTrueBool
+           (* we don't remove the log calls corresponding to source code
```

Comment:
This is important for logging but is placed at a rather odd position. Could we relocate this bit of logic in the Scopelang -> Dcalc translation?